### DISCUSSION TITLE
Sunrise Sunset Times

### DISCUSSION THREAD

**Discussion by Yeshia** [weather-gov/api#359] [Q&A]
Do any of the API endpoints provide the Sunset and Sunrise time for that day?

Thanks

**Comment by netbymatt** [comment]
No they don't. I was looking for this some time ago when I was putting together my app. I was working in Javascript so I used https://github.com/mourner/suncalc to get that data. It's a fairly easy integration since you're either searching by Lat/Lon, or can get the Lat/Lon from one of the API calls and then pass that to the suncalc library to get sunrise and sunset.

**Comment by StephenClouse** [comment]
Correct, it's not something we currently expose. We do actually calculate it internally, because it's needed for some metadata bits on the forecast and observation endpoints, but we don't output the actual times anywhere.

I'm honestly not really sure where it would fit. Maybe a whole new endpoint. Suggestions welcome.

**Comment by Yeshia** [comment]
Perhaps in the daily forecast endpoint? Seems like it might make sense there.

**Comment by StephenClouse** [comment]
Maybe, but the forecast isn't actually daily, it's semi-daily (12-hour periods, 6-6), and sunrise/sunset don't always fit neatly into those boxes, especially as we get around the solstices.

And don't even get me started on Alaska.

Still, it's something we'll try to think about.

**Reply by Shom770** [reply]
The sunrise/sunset could be used for isDaytime in hourly forecasts, because it relies on UTC

**Comment by katilley** [comment]
Has this been turned into a formal feature request? I'd like to see this additional functionality as well. Where ever it goes it would be nice to have the sunrise/sunset for the days available in the forecast.

**Comment by TlonUqbar** [comment]
Hi there,

So I take it that this has not been added/exposed yet?

I think this could easily be included in the  `/gridpoints/{wfo}/{x},{y}/` endpoint.

It already includes all kinds of other info/data.

It includes an `updateTime` and a `validTimes` field, so why not just add a `sunData` field with, the `sunrise` and `sunset` times and a `validDate`, since the `validTimes` includes 7 days then it could be added for 7 days as well, but I have a feeling that most people would only be interested in the first value.
 
```json
{
  "properties": {
    "validTimes": "2025-02-01T14:00:00+00:00/P7DT11H",
    "sunData": {
      "0": { 
        "sunrise": "2025-02-01T06:46:00",
        "sunset": "2025-02-01T17:22:00",
        "validDate": "2025-02-01"
      },
      "1": { 
        "sunrise": "2025-02-01T06:50:00",
        "sunset": "2025-02-01T17:26:00",
        "validDate": "2025-02-02"
      }
    }
  }
}
```

I think this is the best place for it since there is all sorts of other data there, and this would be similar to how all the other data is exposed. Just an idea.


UPDATE:  I was just thinking that `astronomicalData` would be a better key and it could also include other data like moonrise and moonset.


```json
{
  "properties": {
    "validTimes": "2025-02-01T14:00:00+00:00/P7DT11H",
    "astronomicalData": {
      "0": { 
        "sunrise": "2025-02-01T06:46:00",
        "sunset": "2025-02-01T17:22:00",
        "validDate": "2025-02-01",
        "moonrise": "2025-02-01T08:52:00",
        "moonset": "2025-02-01T21:13:00"
      },
      "1": { 
        "sunrise": "2025-02-01T06:50:00",
        "sunset": "2025-02-01T17:26:00",
        "validDate": "2025-02-02",
        "moonrise": "2025-02-01T09:21:00",
        "moonset": "2025-02-01T22:21:00"
      }
    }
  }
}
```

**Reply by zalegrala** [reply]
I would also like to see this in the payload.  A separate `astronomicalData` could make sense for this, which could be appended to later if future use cases dictate.  Currently, the sunrise/sunset is the only missing piece preventing me from implementing the API for my use case.

**Reply by zalegrala** [reply]
What I didn't realize when I posted this is that the value is relatively straight forward to calculate based on lat/long.  I started using github.com/nathan-osman/go-sunrise to perform the calculation and resolves my desire to have this in the API, since it would be more work to call an API than to calculate locally.

**Comment by Dygear** [comment]
This is what I wrote up many years ago in PHP. It doesn't have a return type on the classes, so you know it's been a REALLY long time since I've touched this code.

```PHP
namespace Whatever;
use \DateTimeInterface as DateTimeInterface;

class DateTime extends \DateTime
{
    /**
    * Sunrise
    *
    * Returns time of sunrise for a given day and location.
    *
    * @param DateTime $timestamp If not set assumes this instance, otherwise must be a DateTime object.
    * @param float $latitude Defaults to North, pass in a negative value for South. See also: date.default_latitude.
    * @param float $longitude Defaults to East, pass in a negative value for West. See also: date.default_longitude.
    * @param float $zenith Default: date.sunrise_zenith
    * @param float $gmt_offset Specified in hours.
    * @return Returns the sunrise time in a specified format on success or FALSE on failure.
    */
    public function sunrise(DateTimeInterface $timestamp = null, float $latitude = null, float $longitude = null, float $zenith = null, float $gmt_offset = null)
    {
        $timestamp = ($timestamp) ? $timestamp->format('U') : $this->format('U');
        $latitude = ($latitude) ?: ini_get("date.default_latitude");
        $longitude = ($longitude) ?: ini_get("date.default_longitude");
        $zenith = ($zenith) ?: ini_get("date.sunrise_zenith");
        $return = date_sunrise($timestamp, SUNFUNCS_RET_TIMESTAMP, $latitude, $longitude, $zenith, $gmt_offset);

        return ($return) ? (new DateTime())->setTimestamp($return) : false;
    }

    /**
    * Sunset
    *
    * Returns time of sunset for a given day and location.
    *
    * @param DateTime $timestamp If not set assumes this instance, otherwise must be a DateTime object.
    * @param float $latitude Defaults to North, pass in a negative value for South. See also: date.default_latitude.
    * @param float $longitude Defaults to East, pass in a negative value for West. See also: date.default_longitude.
    * @param float $zenith Default: date.sunrise_zenith
    * @param float $gmt_offset Specified in hours.
    * @return Returns the sunset time in a specified format on success or FALSE on failure.
    */
    public function sunset(DateTimeInterface $timestamp = null, float $latitude = null, float $longitude = null, float $zenith = null, float $gmt_offset = null)
    {
        $timestamp = ($timestamp) ? $timestamp->format('U') : $this->format('U');
        $latitude = ($latitude) ?: ini_get("date.default_latitude");
        $longitude = ($longitude) ?: ini_get("date.default_longitude");
        $zenith = ($zenith) ?: ini_get("date.sunrise_zenith");
        $return = date_sunset($timestamp, SUNFUNCS_RET_TIMESTAMP, $latitude, $longitude, $zenith, $gmt_offset);

        return ($return) ? (new DateTime())->setTimestamp($return) : false;
    }

    /**
    * Sun Info
    *
    * Returns an array with information about sunset / sunrise and twilight begin / end
    *
    * @param DateTime $timestamp If not set assumes this instance, otherwise must be a DateTime object.
    * @param float $latitude Defaults to North, pass in a negative value for South. See also: date.default_latitude.
    * @param float $longitude Defaults to East, pass in a negative value for West. See also: date.default_longitude.
    * @return Returns array on success or FALSE on failure.
    */
    public function sunInfo(DateTimeInterface $timestamp = null, float $latitude = null, float $longitude = null)
    {
        $timestamp = ($timestamp) ? $timestamp->format('U') : $this->format('U');
        $latitude = ($latitude) ?: ini_get("date.default_latitude");
        $longitude = ($longitude) ?: ini_get("date.default_longitude");

        return date_sun_info($timestamp, $latitude, $longitude);
    }
}
```

**Comment by rr0ss0rr** [comment]
I'll plus one, I currently retrieve that from `api.sunrise-sunset.org`

**Comment by greg-does-weather** [comment]
The `/gridpoints/{wfo}/{x},{y}` endpoint provides data in time ranges whereas sunrise and sunset are instants, so that one doesn't make as much immediate sense to me. And to Stephen's comment 5 years ago, the daily forecast isn't really daily. Naively, I think of sunrise and sunset as being properties of a latitude, so the `/points/{lat},{lon}` endpoint clicks for me.

In keeping with the spirit of the `/gridpoints/{wfo}/{x},{y}` endpoint, though, there could be something like (similar to what @Shom770 suggested):

```json
{
  "properties": {
    "isDaylight": {
      "uom": "boolean",
      "values": [
        {
          "validTime": "2025-12-05T12:35:00+00:00/PT8H56M",
          "value": true
        },
        {
          "validTime": "2025-12-05T21:31:00+00:00/PT15H5M",
          "value": false
        }
      ]
    }
  }
}
```

That's not necessarily friendly to parse, though. On the flip side, it tells you not only sunrise and sunset, but also day and night durations, so that's kinda cool.

**Comment by greg-does-weather** [comment]
Sunrise/sunset/etc are available at preview-api.weather.gov. It lives on the `points/` endpoint. For example:

https://preview-api.weather.gov/points/35.683,-101.684

Note that this only includes the solar data for the current day, local to the point requested. Unsure if we'll eventually expand that in some way to include this information for future days and how that would work, but feedback welcome. :)

**Reply by Dygear** [reply]
That's awesome! Thank you!

**Reply by kollymars** [reply]
This was pushed to production today and is on api.weather.gov.  Please use that for operational purposes.
